ark/phenotyping/som_utils.py

```diff
@@ -80,46 +91,109 @@ def create_pixel_matrix(img_xr, seg_labels, fovs=None, channels=None, blur_facto
+    while True:
+        # convert from byte string
+        output = process.stdout.readline().decode('utf-8')
 
+        # if the output is nothing and the process is done, break
+        if process.poll() is not None:
+            break
+        if output:
+            print(output.strip())
 
-def cluster_pixels(chan_list, base_dir,
-                   pixel_pre_name='pixel_mat_preprocessed.csv',
-                   pixel_cluster_name='pixel_mat_clustered.csv'):
-    """Run the FlowSOM training on the pixel data.
 
-    Saves results to pixel_mat_clustered.csv in base_dir.
-    Usage: Rscript som_runner.R {path_to_pixel_matrix} {chan_list_comma_separated} {save_path}
+def cluster_pixels(fovs, channels, base_dir, pre_name='pixel_mat_preprocessed.hdf5',
+                   weights_name='weights.hdf5', cluster_name='pixel_mat_clustered.hdf5'):
+    """Uses trained weights to assign cluster labels on full pixel data
+
+    Saves data with cluster labels to base_dir/pixel_mat_clustered.hdf5
 
     Args:
-        chan_list (list):
+        fovs (list):
+            The list of fovs to subset on
+        channels (list):
             The list of markers to subset on
         base_dir (str):
-            The path to the directory to save the clustered pixel matrix in
-        pixel_pre_name (str):
-            The name of the preprocessed file name, default to pixel_mat_preprocessed.csv
-        pixel_cluster_name (str):
-            The name of the file to write the clustered csv to, default to pixel_mat_clustered.csv
+            The path to the data directory
+        pre_name (str):
+            The name of the preprocessed file name, defaults to pixel_mat_preprocessed.hdf5
+        weights_name (str):
+            The name of the weights file, defaults to weights.hdf5
+        cluster_name (str):
+            The name of the file to write the clustered csv to, default to pixel_mat_clustered.hdf5
     """
 
-    # set the paths to the preprocessed matrix and clustered matrix
-    preprocessed_path = os.path.join(base_dir, pixel_pre_name)
-    clustered_path = os.path.join(base_dir, pixel_cluster_name)
+    preprocessed_path = os.path.join(base_dir, pre_name)
+    weights_path = os.path.join(base_dir, weights_name)
+    clustered_path = os.path.join(base_dir, cluster_name)
 
     # if path to the preprocessed file does not exist
     if not os.path.exists(preprocessed_path):
-        raise FileNotFoundError('Pixel preprocessed path does not exist')
+        raise FileNotFoundError('Pixel preprocessed HDF5 %s does not exist in base_dir %s' %
+                                (pre_name, base_dir))
+
+    # if path to the weights file does not exist
+    if not os.path.exists(weights_path):
+        raise FileNotFoundError('Weights HDF5 %s does not exist in base_dir %s' %
+                                (weights_name, base_dir))
+
+    process_args = ['Rscript', '/som_cluster.R', ','.join(fovs), ','.join(channels),
```

Comment:
Right now, there's no way for us to check if the list of channels provided here is the same list of channels provided during training. I think we'll need to use a slightly more complex data structure to store the weights, one that also includes the list of channel names used to create that weights file. That way, we can validate that the provided channels match the ones used during training.

Reply:
So the good news is that the weights file actually gets saved with the column names. We can easily verify this in `som_cluster.R` and terminate the program if it fails this test.

Reply:
Perfect. Feather to the rescue!